documentation/projects/proposals/document_all_media_properties/20230426-implementation_plan_documenting_media_properties.md

```diff
@@ -0,0 +1,137 @@
+separate markdown file as it is easier to read, write, and lint.
+
+The catalog has a DAG documentation generator that extracts the docstrings from
+the DAGs and creates a page with the documentation. We can use the same approach
+to create a documentation page for the media properties. The DAG doc generator
+runs the code inside docker to extract DAG information. In this project, we can
+simplify and use the `ast` module to parse static code and extract the
+docstrings to make the checks faster. This will, however, mean that some of the
+settings will not be picked up and will have to be updated during the
+generation. One example of this is the `nullable` field in the `Column` class:
+it falls back to the value of `not nullable` during the object initiation.
+
+## Outlined Steps
+
+### `just` script
+
+The script parses the SQL DDL files, Python files and markdown files to generate
+a markdown page with a table and long-form notes. This script can `diff` with
+the existing documentation page (created in step 4 below) for the `precommit`
+check and post a note about how to update the docs if differences are detected.
+
+### 1. SQL DDL parsing
+
+We will use the names of the properties as they are in the upstream database, in
+the same order as they are in the `image_view`/`audio_view` materialized views.
+These views are selected over the main `image` and `audio` tables because they
+are used by the API during data refresh and contain a complete list of
+properties.
+
+A Python script is written to extract the information from the SQL files into a
+mapping with the name of the media property and the following values: SQL
+datatype, and SQL constraints. More data will be added to this mapping in the
+further steps.
+
+### 2. Parsing of the Python code
+
+The `Column` class and its child classes describe the validations we use to
+write the data collected by provider scripts. The `add_item` method has
+docstrings with short descriptions of what each property is. The information
+from these items will be added into the mapping from step 1.
+
+### 3. Adding the long-form information from the markdown file
+
+The markdown file will be parsed and split into the preamble, postamble (???)
+and the information about each data property. This information would be added to
+the `description` in the property mapping.
+
+### 4. Writing the markdown file
+
+The mapping will be used to create a new markdown file: the Preamble, the tables
+for Audio and Image, the long-form document from step 3.
+
+### Information to include in `media_properties.md` for each property
+
+Media types: Image, Audio Short description Names used in the provider scripts,
+if different from the database name Shape of the object (if it is not a simple
+type) Selection criteria Normalization and validation performed in the
+`MediaStore` class and the relevant `Column` Inconsistencies in the database
+data See sample for `identifier` and `tags` in `sample_media_properties.md`
+
```

Comment:
Was this intended to be a list? It's a little hard to follow in the rendered markdown with it altogether like this 😅